### PR TITLE
EOS-14116:Cleanup cortxfs_pmds_fini path

### PR DESCRIPTION
**Pull request by AtitaShirwaikar**
#[EOS-14116] CORTX-NFS : Observed segmentation fault after creating export and restarting ganesha service with firewall enabled.
## Solution Overview
Change description:
               Clean up pmds context only if pNFS role is set to MDS or BOTH

- [x] **Compilation:** _This patch does not break compilation_
- [x] **Merge conflicts:** _This patch has been squashed and re-based, it can be merged using fast-forward merge_
- [] **Code review:** _All discussions have been resolved_
- [x] **Sanity Testing:** _All Unit tests are passing and able to do mount and io operations works from NFS client_
- [x] **Documentation:** _This patch and merge request have up to date description_
- [x] **Unit Testing and debugging:** _Tested on single node VM_

            Unit test (on LABVM):
1)Setup abrt cli as per instructions here.NOTE: Without abrt cli i had not noticed the segfault.Ganesha services were up and running and no error was observed in ganesha logs.
https://teams.microsoft.com/_#/txt/viewer/teams/https:~2F~2Fseagatetechnology.sharepoint.com~2Fsites~2FCORTX~2FShared%20Documents~2FEOS-NFS~2FSetup_for_ganesha_coredump.txt?threadId=19:8387aec0f02040debb465d4f8797700f@thread.tacv2&messageId=1600946875393&baseUrl=https:~2F~2Fseagatetechnology.sharepoint.com~2Fsites~2FCORTX&fileId=f92a0740-040f-46fa-bc1b-86b254b146f0&ctx=chiclet&viewerAction=view
2)cortxfscli fs create rtfs2
3)cortxfscli endpoint create rtfs2 proto=nfs,secType=sys,Filesystem_id=192.1,client=1,clients=*,Squash=no_root_squash,access_type=RW,protocols=4,pnfs_enabled=false
4)systemctl restart nfs-ganesha
5)abrt-cli ls
No core dumps should be seen here

### CODE REVIEW

File: src/FSAL/FSAL_CORTXFS/mds.c [hunk 1]
```diff
@@ -485,29 +485,30 @@ int cortxfs_pmds_ini(struct kvsfs_fsal_module *kvsfs,
...
+		cortxfs_pmds_cleanup(kvsfs->mds_ctx->mds);
 
+		rc = pthread_mutex_unlock(&kvsfs->mds_ctx->mds_mutex);
+		if (rc != 0) {
+			LogCrit(COMPONENT_PNFS, "pthread_mutex_unlock failed: %x", rc);
+			goto out;
+		}	
+	
+  	}
 	rc = pthread_mutex_destroy(&kvsfs->mds_ctx->mds_mutex);
```
Comment (pratyush-seagate) [inline src/FSAL/FSAL_CORTXFS/mds.c:507]: Pls put this mutex destroy code inside the if block from line number 490. Rest looks good.

**Review by pratyush-seagate — APPROVED** [review]
LGTM, +1

File: src/FSAL/FSAL_CORTXFS/mds.c [hunk 2]
```diff
@@ -485,29 +484,31 @@ int cortxfs_pmds_ini(struct kvsfs_fsal_module *kvsfs,
...
-	rc = pthread_mutex_unlock(&kvsfs->mds_ctx->mds_mutex);
-	if (rc != 0) {
-		LogCrit(COMPONENT_PNFS, "pthread_mutex_unlock failed: %x", rc);
-		goto out;
-	}
+		rc = pthread_mutex_unlock(&kvsfs->mds_ctx->mds_mutex);
+		if (rc != 0) {
+			LogCrit(COMPONENT_PNFS, "pthread_mutex_unlock failed: %x", rc);
+			goto out;
+		}
```
Comment (sachinpunadikar) [inline src/FSAL/FSAL_CORTXFS/mds.c:503]: Remove the tab after }

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
remove the extra tab

**Comment by sachinpunadikar** [comment]
I was thinking what if we add this check (kvsfs->mds_ctx->pnfs_role > CORTXFS_PNFS_DS) in the caller function kvsfs_unload ?

**Comment by AtitaShirwaikar** [comment]
> I was thinking what if we add this check (kvsfs->mds_ctx->pnfs_role > CORTXFS_PNFS_DS) in the caller function kvsfs_unload ?

mds_ctx  structure contains pnfs_role.This struct needs to be freed in cortxfs_pmds_fini irrespective of the role.Maybe it will be more appropriate if i change mds_ctx name to pnfs_ctx name.That will cause less confusion.

**Review by sachinpunadikar — APPROVED** [review]
+1

**Pull request merged**